Merge common parts of FutureNCCL into at::ivalue::Future

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #49014 Drop FutureNCCL in favor of vanilla CUDAFuture
* #48789 Make CUDAFuture remember and restore current device in callback
* #48840 Remove DataPtr extractor from CUDAFuture
* #48788 Cache the DataPtrs in CUDAFuture
* #48506 Split out reusable CUDAFuture from FutureNCCL
* **#48505 Merge common parts of FutureNCCL into at::ivalue::Future**
* #48504 Split FutureNCCL's CUDA-specific parts from generic future logic
* #48502 Support wider range of types in FutureNCCL
* #48501 Don't store device indices separately on FutureNCCL
* #48500 Add multi-GPU support to FutureNCCL
* #48563 Fix FutureNCCL not recording dataptrs with caching alloc in wait()
* #48503 Fix FutureNCCL's completed() disagreeing with wait()
* #48499 Record CUDA events for "follow-up" FutureNCCL inside markCompleted
* #48498 Use fresh stream from pool for each FutureNCCL callback
* #48497 Make FutureNCCL record events in current stream
* #48496 Have FutureNCCL record streams w/ allocator in addCallback
* #48562 Add some safeguards to FutureNCCL
* #48495 Remove NCCL dependency from PythonFutureWrapper
* #48561 Avoid using FutureNCCL before it's ready

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

lw committed:
Merge common parts of FutureNCCL into at::ivalue::Future

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -398,22 +405,18 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
       std::function<IValue(void)> callback,
       TypePtr type) {
-    auto fut = c10::make_intrusive<Future>(type);
-    // Cannot move capture std::function in lambda, because it cannot deduce
-    // the template type for std::function. Hence use std::bind to explicitly
-    // specify types.
```
I was curious to see what the reason for this problem was so I tried to undo this fix... and it seems to work? Maybe the comment was outdated?

lw committed:
Update on "Merge common parts of FutureNCCL into at::ivalue::Future"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -479,6 +482,17 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
   }
 
+ protected:
+  virtual c10::intrusive_ptr<Future> createInstance(at::TypePtr type) {
+    return c10::make_intrusive<Future>(type);
+  }
+
+  virtual void postMarkCompletedHook(const at::IValue& value) {}
+
+  virtual std::function<void(void)> wrapCallback(std::function<void(void)> callback) { return callback; }
+
+  virtual void postWaitHook() {}
```
Shall we add some comments to these functions and explain what derived classes need to do when implementing them?

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -213,67 +213,22 @@ class ProcessGroupNCCL : public ProcessGroup {
-      if (error_) {
-        throw *error_;
+    void setDataPtrExtractor(DataPtrExtractor data_ptr_extractor) override {
+      // To avoid races with other threads that may be using the extractor, we
```
do we expect this function to be called multiple times? If no, do we need a lock + assert?

lw replied:
Yes, indeed I was being sloppy here, thanks for calling me out. It can be called multiple times: once when the future is constructed (if it's a "child" future created by the `then()` method), and then one more time by PythonFutureWrapper for each time we call `then()` on that future (in order to create a "grand-child" future). Now that I think of it this data race was probably there before, but we should still fix it.

I'm still hoping we can come up with a better idea for this whole DataPtrExtractor, or get rid of it entirely if we can get a JIT helper that can do it for us. If neither of that happens, I'll add a lock here.

mrshenli reviewed (approved):
LGTM. Please check with @wanchaol regarding whether adding cbs to ivalue Future is OK.

lw committed:
Update on "Merge common parts of FutureNCCL into at::ivalue::Future"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]

lw committed:
Update on "Merge common parts of FutureNCCL into at::ivalue::Future"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]

lw committed:
Update on "Merge common parts of FutureNCCL into at::ivalue::Future"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]

lw committed:
Update on "Merge common parts of FutureNCCL into at::ivalue::Future"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]

lw commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -125,7 +125,7 @@ def compute_q(fut):
             dist.all_reduce(q, group=group_to_use, async_op=True)
             .get_future()
-            .value()[0]
+            .wait()[0]
```
This is needed because, in the PythonFutureWrapper, calling wait() also returns a value(), whereas calling value() does _not_ wait. This was already true on the CPU side (i.e., value() fails if the future isn't complete), and now it's also true for the GPU part (we don't sync streams when calling value()).

This was a discrepancy between ivalue::Future and FutureNCCL which I think we should fix.

The value() method is thus supposed to only be used to retrieve the value once we know it's already been correctly waited/synchronized, for example within a callback.

lw committed:
Update on "Merge common parts of FutureNCCL into at::ivalue::Future"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]

wanchaol commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -290,18 +290,22 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
       finished_cv_.wait(lock);
     }
+
+    if (!eptr_) {
```
shall we use `hasError` instead?

lw replied:
I don't think we can use it as-is, since it tries to acquire the mutex, and this method already has it, and that would deadlock. If you think it's worth it I can add a `hasErrorInternal` method that doesn't acquire the mutex.

wanchaol commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -479,6 +484,17 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
+    return c10::make_intrusive<Future>(type);
+  }
+
+  virtual void postMarkCompletedHook(const at::IValue& value) {}
```
can you briefly document how these there apis are used, probably only FutureNCCL is using them now, but if there're other future derived type there, might be a good reference.

wanchaol reviewed (approved):
looks good to me.

lw committed:
Update on "Merge common parts of FutureNCCL into at::ivalue::Future"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]

lw committed:
Update on "Merge common parts of FutureNCCL into at::ivalue::Future"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]

lw committed:
Update on "Merge common parts of FutureNCCL into at::ivalue::Future"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL isn't just adding CUDA support to ivalue::Future, it's also reimplementing a lot of the latter's logic (by overriding plenty of its methods). That's brittle, as whenever a new method is added to ivalue::Future there's a risk of forgetting to add it to FutureNCCL, and in such a case calling this method on FutureNCCL would defer to the base class and give inconsistent results (e.g., future not being completed when it actually is). This _is already happening_, for example with the waitAndThrow or hasError, which are not implemented by FutureNCCL. In addition, this creates duplication between the two classes, which could lead to inconsistencies of behavior, bugs, missing features, ...

The best solution would be to keep the core future logic in ivalue::Future, and have _only_ the CUDA additions in FutureNCCL. That's what we're going to do, in two steps. In the previous commit, I split the CUDA features into separate hooks, which are called by FutureNCCL's other methods. In this commit, I'm removing these latter methods, and invoke the hooks directly from ivalue::Future.

Differential Revision: [D25180535](https://our.internmc.facebook.com/intern/diff/D25180535/)

[ghstack-poisoned]
